### ISSUE TITLE
[HOLD for payment 2025-01-30] [$250] Refactor SMS Delivery Failure sign in flow

### ISSUE THREAD

**Issue by carlosmiceli** [Expensify/App#53980]
We need to refactor the existing flow blacklisting phone numbers (see steps in screenshots). This is triggered when the user tries to sign in with a phone number that failed to receive Twilio SMS. This is a follow up to this [issue](https://github.com/Expensify/App/issues/52086).

---

When calling `BeginSignIn`, if `response[smsDeliveryFailureStatus]` is not `null`, then it should display the following message:

```
"We've been unable to deliver SMS messages to [PHONE-NUMBER], so we've suspended it for 24 hours. Please try validating your number:"
```

There should also be a "Validate" button that calls `ResetSMSDeliveryFailureStatus`. 

If `response[hasSMSDeliveryFailure]`is set to`true`, we should display `Validation failed because it hasn't been 24 hours since your last attempt. ${response.message}` and a "Got it" button that sends the user back to the beginning of the sign in flow.

If `response[hasSMSDeliveryFailure]`is set to`false`, we should display `Your number has been validated! Click below to send a new magic sign-in code.` and a "Send" button that calls `BeginSignIn` as usual.

Once this is complete and merged I'll do the internal QA with the BE.
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021867038175379475893</li>
        <li>Upwork Job ID: 1867038175379475893</li>
        <li>Last Price Increase: 2024-12-12</li>
<li>Automatic offers: </li>
<ul>
<li>nyomanjyotisa | Contributor | 105327292</li>
</ul></li>
    </ul>
</details>

**Comment by hungvu193** [comment]
I'll review proposal tomorrow.

**Comment by carlosmiceli** [comment]
@hungvu193 thank you, this is kinda urgent but tomorrow works! 👍

**Comment by hungvu193** [comment]
@nyomanjyotisa 's proposal looks good to me!

🎀 👀 🎀

**Comment by carlosmiceli** [comment]
ResetSMSDeliveryFailure has not changed, it's the same structure as before.

BeginSignIn's response currently includes:

```
"smsDeliveryFailureStatus": {
    "hasSMSDeliveryFailure": true / false
},
```

This will change to:
```
"smsDeliveryFailureStatus": null / not null
```

**Comment by hungvu193** [comment]
Draft PR is created.

**Comment by Christinadobrzyn** [comment]
I think we moved the PR to this one—https://github.com/Expensify/App/pull/54400—which is still under review. So payment day isn't today. 

Is that correct @hungvu193 @nyomanjyotisa?

**Comment by hungvu193** [comment]
> I think we moved the PR to this one—https://github.com/Expensify/App/pull/54400—which is still under review. So payment day isn't today. 
> 
> 
> 
> Is that correct @hungvu193 @nyomanjyotisa? 

Yeah. PR is still on hold for BE changes. Payment isnt due yet

**Comment by Christinadobrzyn** [comment]
Thanks @hungvu193! I'll add a HOLD to this GH for https://github.com/Expensify/App/pull/54400

**Comment by Christinadobrzyn** [comment]
Thanks @hungvu193! I'll add a HOLD to this GH for https://github.com/Expensify/App/pull/54400

**Comment by Christinadobrzyn** [comment]
I think we can move this to weekly while we wait for https://github.com/Expensify/App/pull/54400?

**Comment by carlosmiceli** [comment]
Found a small bug on the command name and one displaying both email and sms errors, just commented on the PR. After those changes I think this should be ready to go.

**Comment by carlosmiceli** [comment]
PR was merged.

**Comment by carlosmiceli** [comment]
This can be moved to weekly now that we're waiting for QA.

**Comment by Christinadobrzyn** [comment]
@hungvu193 do we need a regression test for this?

**Comment by hungvu193** [comment]
> [@hungvu193](https://github.com/hungvu193) do we need a regression test for this?

I think Yes.

**Comment by hungvu193** [comment]
**Regression Test:**

**Test 1:**
1. Login with account that has `account.smsDeliveryFailureStatus` that equals empty value: (account.smsDeliveryFailureStatus: {})
2. Verify that the `SmsDeliveryFailurePage` is displayed correctly.
3. Click `Validate` button.
4. Verify that the message is displayed correctly
5. Click the "Got it" button
6. Verify that the user is navigated to the beginning of the sign-in flow.

**Test 2:**
1.  Login with account that has `{smsDeliveryFailureStatus: {isReset: true, hasSMSDeliveryFailure: false}`
2. Click the "Validate" button
3. Verify that the message is displayed correctly
4. Click the "Send" button
5. Verify that the BeginSignIn API called
6. Wait for a few seconds and wait BeginSignIn API to be called successfully.
7. Verify that the user was able to continue the sign-in flow and navigate to the magic code request

Do we 👍  or 👎 ?

**Comment by carlosmiceli** [comment]
@hungvu193 A bit confused, what is each test supposed to validate exactly? There should be 3 tests, right?:
- Null `smsDeliveryFailureStatus` (nothing happens, sign in fine). This is test 1.
- Not null `smsDeliveryFailureStatus`, which means testing for resetting and failing ro reset (tests 2 and 3).

**Comment by hungvu193** [comment]
> [@hungvu193](https://github.com/hungvu193) A bit confused, what is each test supposed to validate exactly? There should be 3 tests, right?:
> 
> * Null `smsDeliveryFailureStatus` (nothing happens, sign in fine). This is test 1.
> * Not null `smsDeliveryFailureStatus`, which means testing for resetting and failing ro reset (tests 2 and 3).

My bad, I updated the tests with 3 tests for 3 cases

**Comment by hungvu193** [comment]
I updated the [tests](https://github.com/Expensify/App/issues/53980#issuecomment-2630097211) again to make it more clear

**Comment by carlosmiceli** [comment]
Great, tests look good now!

**Comment by Christinadobrzyn** [comment]
Just catching up, are we ready to pay this out or are we still working on anything? 

I'll make sure to create a regression test for these [tests](https://github.com/Expensify/App/issues/53980#issuecomment-2630097211)

**Comment by hungvu193** [comment]
> Just catching up, are we ready to pay this out or are we still working on anything? 
> 
> I'll make sure to create a regression test for these [tests](https://github.com/Expensify/App/issues/53980#issuecomment-2630097211)

This is ready for payout already 😁

**Comment by Christinadobrzyn** [comment]
Regression test: https://github.com/Expensify/Expensify/issues/468252

Payment summary 

Contributor: @nyomanjyotisa paid $250 via Upwork (paid https://www.upwork.com/nx/wm/workroom/39243599)
Contributor+: @hungvu193 owed $250 via NewDot

@hungvu193 please make sure to request payment in ND. Closing as complete

**Comment by garrettmknight** [comment]
$250 approved fro @hungvu193
